### ISSUE TITLE
New Effect API to allow for reusable side-effect methods

### ISSUE THREAD

**Issue by bdwain** [raisemarketplace/redux-loop#66]
Copied this comment from #54 

After using redux-loop a bit more in practice, I've started to have some thoughts on how the api could encourage a little more code reuse and simplify a few things.

One issue with forcing effects to return/resolve with redux actions is that the implementations of the side-effects become tied to redux/redux-loop. Suddenly a pretty standard function to make a request, such as this,

```js
import apiClient from './apiClient.js'; //to abstract away fetch details like json parsing and error codes

function fetchItems(userid){
  return apiClient(`/users/${userId}/items`)
      .then(rsp => {
           //process rsp
           return processedRsp;
       }).catch(err => {
            console.log(':(');
            return err;
       });
}
```

can't be used by redux-loop without changing it in a way that makes it unusable by something else (you can no longer reject the promise, and your resolution needs to be a redux action and not just the list of items).

Additionally, even if you wanted it only for redux-loop, it would be impossible (or at least awkward) to use a side-effect to return different types of actions depending on how it is called (I haven't had a need to do this yet, but it's not hard to imagine a case where you'd want it)

What if we allowed side-effects to become completely agnostic to redux-loop/redux in general, and instead return whatever they felt was necessary to complete their purpose (so for async methods, prob would be a promise that resolves or rejects to indicate success).

So instead of defining the actions triggered (or not) by the effect in the method, it could be defined in the Effect object itself.

The new effect objects returned by a reducer might look like this

```js
//actions.js
function itemsLoadedActionCreator(items){
    return {
       type: 'ITEMS_LOADED',
       items: items
    };
}
// ... other action creators 

//reducer.js
return loop(newState, Effect.promise(fetchItems, itemsLoadedActionCreator, failHandler, params...));

Effect.call(writeToLocalStore, syncMethodActionCreator, params...)

//Effects.constant could be implemented as syntactic sugar for 
Effects.call(null, constantAction());

//batch would be basically the same
//I imagine lift would be similar, though I haven't looked into it much yet
```

And now your effects are completely reusable and agnostic to your choice of state management. You could swap out redux-loop or even redux iteself without having to change them.

**Comment by bdwain** [comment]
Additionally, and it could be separate from this, I think it could be useful to allow passing the dispatch method to side-effects. I have a few scenarios where it would be nice if the side effect was able to dispatch an action, without it being the specific action that is the result of the effect.

**Comment by lukewestby** [comment]
Let's leave passing `dispatch` to effects for another discussion. Feel free to open up another issue about it!

So here are my initial thoughts:

I take your point about wanting to keep actual effects separate from the redux-loop API very well. This is something I've started to notice in code at work too - even when we do have reusable effect functions they still need to be wrapped in an entire additional function that can be exported in order to test reducers. Passing the action creators for success and failure has a ton of positive impact on this problem.

- Just exactly as you said, functions that produce Promises can just return a Promise for the thing that they are doing. Action creators can and should be separate. This change can be seen in Elm as well. In 0.17, rather than mapping a `Task` directly onto success and failure actions, you take an ordinary `Task` and "perform" it, turning it into a `Cmd` (the new term for `Effect`, if you haven't been following). It looks more like this now:

```elm
LoadStuff id ->
  ( { model | stuff = Loading }
  , Api.loadStuff id |> Task.perform LoadStuffFailure LoadStuffSuccess
  )

LoadStuffSuccess stuff ->
  ( { model | stuff = Success stuff }
  , Cmd.none
  )

LoadStuffFailure errorMessage ->
  ( { model | stuff = Failure errorMessage }
  , Cmd.none
  )
```

Where `Api.loadStuff` is returning a `Task String (List Thing)` or whatever and the conversion to actions happens later. So, this is a long-winded way of saying I think this is a great idea.

- We no longer need to track if the user actually passed something into `.catch()` and print a confusing error message. We can simply throw if the user doesn't call the effect creator correctly and let them know that they left off the argument for the failure action creator or some such.

-----

So here is my initial vision for an API based on your ideas:

- Let's rename `Effects` to `Cmd` to keep consistency since this is still a port
- Let's create a bunch of `Cmd.from{Something}` functions that enforce their arguments but keep things nicely separated
  - `Cmd.fromPromise(promiseCreator, successActionCreator, failureActionCreator, ...params)`
  - `Cmd.fromFailproofPromise(promiseCreator, successActionCreator, ...params)` - for when you're doing like a timeout or something
  - `Cmd.fromConstant(action)`
  - `Cmd.fromSync(syncEffect, successActionCreator, failureActionCreator, ...params)` - as a replacement for `Effects.call`, will probably need more discussion
  - Maybe `Cmd.fromCallback(nodeStyleFunction, successCreator, failureCreator, ...params)` for convenience?

And then the pre-existing
  - `Cmd.batch(otherCmds)`
  - `Cmd.map(tagger, otherCmd)`
  - `Cmd.none`

Function names are all up for debate and suggestions.

**Comment by bdwain** [comment]
Thanks for getting back so quickly @lukewestby. 

I haven't used Elm before, but I like the idea of keeping it in sync for people who have. A few other thoughts...

- Is the from prefix necessary? If they basically all have it, it might be cleaner to just get rid of it.
- FromSync wouldn't need a fail handler (unless it got called due to an exception being thrown. but idk if that's necessary)
- Is Cmd.map equivalent to Effects.lift? Otherwise i'm not sure what it is.
- Not sure about failproof promise. I feel like it could be combined with promise. Why not just pass nothing for the fail handler.
- I'm confused how the fromCallback would look. How would the action creators get triggered?

I know it goes back to the other issue I posted, but making all of these action creators optional (and in the loop code just say `if actionCreator, dispatch result`) could cut down a lot of unnecessary code. Most of my synchronous cmds don't dispatch actions when done, and having to pass a noop generator to all of them seems like a waste. And either way, it'd be nice not to fill the action history with a bunch of noop actions. I have it implemented with them being optional in a branch [here](https://github.com/bdwain/redux-loop/tree/update-version) (along with the sequence effect we discussed in another issue)

**Comment by lukewestby** [comment]
> Is the from prefix necessary? If they basically all have it, it might be cleaner to just get rid of it.

It isn't, let's leave it off

> FromSync wouldn't need a fail handler (unless it got called due to an exception being thrown. but idk if that's necessary)

(We can refer to this as `call` again) we can leave the failure handler off for now and log something to the console if a `call` effect ends up throwing

> Is Cmd.map equivalent to Effects.lift? Otherwise i'm not sure what it is.

It is, I apologize. Let's leave it as `Cmd.lift`.

> Not sure about failproof promise. I feel like it could be combined with promise. Why not just pass nothing for the fail handler.

Leaving the failure handler out would make it hard or impossible to detect if the thing in the failure handler's position is an action creator or one of the params. I would prefer two separate functions here. Doing async things that never fail has come up several times for us in the past, I'd like for this to be included in the API.

> I'm confused how the fromCallback would look. How would the action creators get triggered?

It can be wrapped in a Promise, something like this:
```js
new Promise((resolve) => {
  nodeStyleFunction((error, result) => {
    if (error) resolve(failureCreator(error))
    else resolve(successCreator(result))
  })
})
```

For effects that don't return actions, what do you think about a totally separate `Cmd` function that will just accept any function followed by its arguments and call it regardless of whether it returns a promise or whatever. I'm not sure what to call it yet, something like `Cmd.arbitrary(effectFunction, ...params)`. This would effectively be equivalent to port commands in Elm.

**Comment by bdwain** [comment]
> It is, I apologize. Let's leave it as Cmd.lift.

Isn't it map in Elm? It'd be nice to have them the same. Though it is confusing to rename it. If you want to get it to have the same name eventually, we could have lift be an alias for map, and put a deprecation notice in for it or something. I don't feel too strongly about the name either way. 

As far as the effects that don't return actions, what about having the exact same types, but namespaced differently. Either `Arbitrary.call, Arbitrary.promise` or maybe `Cmd.Arbitrary.call`. I'm not really sure what the new name would be either. But something like that would be easy to grasp as doing the same thing but not forcing you to dispatch an action. And could probably be implemented as wrappers around the originals. Though it also would only be relevant for a few of the types (No need for Arbitary.batch constant or none)

Everything else you said sounds good to me.

**Comment by lukewestby** [comment]
@bdwain can you share a code snippet as close as possible without revealing IP that demonstrates the use-case for command action creators being optional?

**Comment by bdwain** [comment]
sure.

```js
export function writeAuthUpdatesToStorage(newAuthState){
  localStorage.setItem(AUTH_TOKEN_STORAGE_IDENTIFIER, newAuthState.get('authToken'));
  localStorage.setItem(USER_ID_STORAGE_IDENTIFIER, newAuthState.get('userId'));
  localStorage.setItem(CURRENT_USER_STORAGE_IDENTIFIER, JSON.stringify(newAuthState.get('currentUser')));
}

//reducer.js

if(action.type === 'SET_AUTH_DATA'){
  newState = state.merge(action); //or something like that
  return loop(newState, Effects.call(writeAuthUpdatesToStorage, newState);
}
```

so in my app, we back up auth data to localstorage so that on refresh you are still logged in. so every change should trigger a localstorage update. but after that, i don't have anything else to do.

**Comment by lukewestby** [comment]
cool. so i mentioned before, but want to revisit that this use case and lots of others could be covered by having a single function `Cmd.arbitrary` or `Cmd.port` or whatever that accepts any function followed by its args and just calls it without caring about what it returns or what the outcome is. For this one it might look like

```js
if (action.type === 'SET_AUTH_DATA') {
  newState = state.merge(action.payload) // or whatever
  return loop(newState, Cmd.arbitrary(writeAuthUpdatesToStorage, newState)
}
```

if in the future you wanted to also add some sort of logging you could write an effect function that made a network request like

```js
export function logState(state) {
  fetch('/logs', { method: 'POST', body: JSON.stringify(state) })
}

// reducer.js

if (action.type === 'SET_AUTH_DATA') {
  newState = state.merge(action)
  return loop(
    newState,
    Cmd.batch([
      Cmd.arbitrary(writeAuthUpdatesToStorage, newState),
      Cmd.arbitrary(logState, newState),
    ])
  )
}
```

it doesn't matter of what form the effect is if you don't care about the outcome so I think most things can be covered by just the one `Cmd.arbitrary` function.

**Comment by bdwain** [comment]
That's true. I guess I was thinking that you needed one for each command type but I guess if you're not checking the return value then it's just calling the function and that's it. 

The only question is what about In a batch (or sequence like we talked about in the other issue). Let's say it was an asynchronous port call. How would the batch know not to dispatch all of its actions until after the port completed. Or if it was a sequence and the effect after the port had to wait until it finished.

Maybe the solution is to have a port and an asyncPort that returns a promise that batching knows to wait for?

**Comment by bdwain** [comment]
Also after reading about ports a little I think the name makes sense for this case

**Comment by bdwain** [comment]
any thoughts on this @lukewestby

**Comment by lukewestby** [comment]
I think that all sounds good, I've rethought naming them `port` and `asyncPort`, since those names in Elm are meant to indicate that you're escaping the constraints of the Elm program into JS-land by exiting and entering through a "port". One more thing before moving forward - can you include an example usage where ordering these things is important? I just always like to see the motivation for a feature when I haven't needed it myself.

**Comment by bdwain** [comment]
Cool. Are you going to stick with arbitary for now then?

When you say an example of where ordering is important, do you mean for Effect.Sequence? The example i needed it for was writing to localStorage first, and then doing something that depended on the value in storage. 

> The only question is what about In a batch (or sequence like we talked about in the other issue). Let's say it was an asynchronous port call. How would the batch know not to dispatch all of its actions until after the port completed. Or if it was a sequence and the effect after the port had to wait until it finished.

It looks like you adressed this question with your isPromiseLike check.

**Comment by lukewestby** [comment]
Let's stick with arbitrary. Do you think being able to optionally return a Promise in an `arbitrary` effect and then detecting that will be sufficient for what you need to do?

**Comment by bdwain** [comment]
yep i think that will address my concern. if you batch something synchronous then the batch doesn't need to wait for anything to happen.

**Comment by lukewestby** [comment]
good deal. i think we can close this issue then! 🎉
